flowmap/particles.py

```diff
@@ -0,0 +1,121 @@
+import pandas as pd
+from tvtk.api import tvtk
+from tvtk.common import configure_input, configure_source_data, configure_input_data
+import vtk
+import geojson
+import pyproj
+
+def make_tracer_pipeline(polydata, seed):
+    # create elements of the pipeline
+    tracer = tvtk.StreamTracer()
+
+
+    # # # You can compute up to 1km per timestep (8km/hr for )
+
+    n = 200  # max number of steps
+    l = 1000
+
+
+    # maximum 1km
+    tracer.maximum_propagation = l
+    # # # In m
+    tracer.integration_step_unit = vtk.vtkStreamTracer.LENGTH_UNIT
+    # # # Minimum 5 per step
+    tracer.minimum_integration_step = (l/n)
+    # # # Maximum 100m per step
+    tracer.maximum_integration_step = 10*(l/n)
+    # # # Maximum 200 steps
+    tracer.maximum_number_of_steps = n
+    # # # Maximum error 1cm
+    tracer.maximum_error = 1e-2
+    # # # We use a path integration. You could argue that you need a
+    # # # particle tracking algorithm that matches the numerical grid
+    # # # (in our case edge velocities
+    # # # and integration over a cell instead of over a line)
+    tracer.integrator_type = 'runge_kutta45'
+
+    tracer.debug = True
+
+    cell2point = tvtk.CellDataToPointData()
+    # setup the pipeline
+    configure_input(cell2point, polydata)
+    configure_input(tracer, cell2point)
+    configure_source_data(tracer, seed)
+    return tracer
+
+def extract_points(tracer):
+    points = tracer.output.points.to_array()
+    points_data = {
+        "point_idx": np.arange(points.shape[0]),
+        "x": points[:, 0],
+        "y": points[:, 1],
+        "z": points[:, 2]
+    }
+
+    for i in range(tracer.output.point_data.number_of_arrays):
+        name = tracer.output.point_data.get_array_name(i)
+        arr = tracer.output.point_data.get_array(i).to_array()
+        points_data[name] = list(arr)
+    return pd.DataFrame(points_data)
+
+def extract_lines(tracer):
+    """convert the streamlines to a data frame"""
+    data = {}
+
+    points = tracer.output.points.to_array()
+
+    lines = tracer.output.lines.to_array()
+
+    for i in range(tracer.output.cell_data.number_of_arrays):
+        arr = tracer.output.cell_data.get_array(i).to_array()
+        name = tracer.output.cell_data.get_array_name(i)
+        data[name] = arr
+    start = 0
+    line_segments = []
+    point_idx = []
+    for i in range(tracer.output.lines.number_of_cells):
+        """loop over al lines"""
+        n = lines[start]
+        idx = lines[(start+1):(start+n+1)]
+        line = points[idx]
+        line_segments.append(line)
+        point_idx.append(idx)
+        start += (n + 1)
+    data['line'] = line_segments
+    data['points'] = point_idx
+    return pd.DataFrame(data)
+
+def make_particles(polydata, n=100, mode='random'):
+    points = polydata.points.to_array()
+    xmax, ymax, zmax = points.max(axis=0)
+    xmin, ymin, zmin = points.min(axis=0)
+    # take the srt of n
+    n = np.round(np.sqrt(n)).astype('int')
+    if mode != 'random':
```

Comment:
Make it a boolean? Random false or true --> otherwise try/catch error if people implement other modes?

Reply:
explicit error on unsupported modes. Now supports 2 modes. A more uniform like https://bl.ocks.org/mbostock/b17e0b2aa8b2d50de465 should look better.